finagle-http/src/main/scala/com/twitter/finagle/http/HeaderMap.scala

```diff
@@ -52,67 +54,88 @@ abstract class HeaderMap
   def += (kv: (String, Date)): HeaderMap =
     += ((kv._1, HeaderMap.format(kv._2)))
 
-  override def empty: HeaderMap = new MapHeaderMap(mutable.Map.empty)
+  override def empty: HeaderMap = MapHeaderMap()
 }
 
+private[finagle] case class HeaderValuePair(header: String, value: String) {
+  val canonicalName = HeaderValuePair.canonicalName(header)
+}
+
+private[finagle] object HeaderValuePair {
+  def canonicalName(header: String) = header.toLowerCase
+}
 
 /** Mutable-Map-backed [[HeaderMap]] */
-class MapHeaderMap(underlying: mutable.Map[String, Seq[String]]) extends HeaderMap {
+class MapHeaderMap extends HeaderMap {
+
+  private val underlying = mutable.Map.empty[String, List[HeaderValuePair]]
+
+  // This is here to provide backward compatibility
+  def this(underlying: mutable.Map[String, Seq[String]]) {
+    this
+    for ((k, vs) <- underlying; v <- vs) yield{
+      add(k, v)
+    }
+  }
 
   def getAll(key: String): Iterable[String] =
-    underlying.getOrElse(key, Nil)
+    underlying.getOrElse(HeaderValuePair.canonicalName(key), Nil).map(_.value)
 
   def add(k: String, v: String): MapHeaderMap = {
-    underlying(k) = underlying.getOrElse(k, Nil) :+ v
+    val t = HeaderValuePair(k, v)
+    underlying(t.canonicalName) = underlying.getOrElse(t.canonicalName, Nil) :+ t
     this
   }
 
   def set(key: String, value: String): MapHeaderMap = {
-    underlying.retain { case (a, _) => !a.equalsIgnoreCase(key) }
-    underlying(key) = Seq(value)
+    val t = HeaderValuePair(key, value)
+    underlying(t.canonicalName) = List(t)
     this
   }
 
   // For Map/MapLike
   def get(key: String): Option[String] = {
-    underlying.find { case (k, v) => k.equalsIgnoreCase(key) }.flatMap { _._2.headOption }
+    getAll(key).headOption
   }
 
   // For Map/MapLike
   def iterator: Iterator[(String, String)] = {
-    for ((k, vs) <- underlying.iterator; v <- vs) yield
-      (k, v)
+    for ((_, vs) <- underlying.iterator; v <- vs) yield
+      (v.header, v.value)
   }
 
   // For Map/MapLike
-  def += (kv: (String, String)): MapHeaderMap.this.type = {
-    underlying(kv._1) = Seq(kv._2)
+  def +=(kv: (String, String)): MapHeaderMap.this.type = {
+    val t = HeaderValuePair(kv._1, kv._2)
+
+    // this slightly complicated logic is here to be backward compatible
+    underlying(t.canonicalName) =
+      underlying.getOrElse(t.canonicalName, Nil)
+        .filterNot(_.header == t.header) :+ t
     this
   }
 
   // For Map/MapLike
   def -= (key: String): MapHeaderMap.this.type = {
-    underlying.retain { case (a, b) => !a.equalsIgnoreCase(key) }
+    underlying.remove(HeaderValuePair.canonicalName(key))
     this
   }
 
   override def keys: Iterable[String] =
```

Comment:
do you think we should document that these will return the keys in a normalized form (here and below)?